### PR TITLE
 Calling .Buffer() on the scanner to permit larger documents in NewDocumentDecoder

### LINKED ISSUE #89016: NewDocumentDecoder fails with bufio.Scanner: token too long
**What happened**:

When calling `NewDocumentDecoder` the token size for the buffer is the default which is NOT changable. Larger documents (which is likely when using a yaml separator) causes a `bufio.Scanner: token too long`

This was uncovered using kudo... the issue is https://github.com/kudobuilder/kudo/issues/1397

**What you expected to happen**:

1. size limitation should be documented if they exist
2. token size should be calling code managed

**How to reproduce it (as minimally and precisely as possible)**:

Call `NewDocumentDecoder` with a doc larger than token size

**Anything else we need to know?**:

The fixing PR https://github.com/kudobuilder/kudo/pull/1400 provides more detail if needed.

Code concern is: https://github.com/kubernetes/apimachinery/blob/release-1.6/pkg/util/yaml/decoder.go#L94

duplicate of https://github.com/kubernetes/apimachinery/issues/89 which I will now close

**Environment**:
- Kubernetes version (use `kubectl version`):  1.17
- Cloud provider or hardware configuration: 
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by 0xMH**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Adding maxCapacity to NewDocumentDecoder to set the Buffer size to 5M to permit larger documents fixes #89016

**Which issue(s) this PR fixes**:
Fixes #89016

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go [hunk 1]
```diff
@@ -92,6 +92,10 @@ type YAMLDecoder struct {
 // the caller in framing the chunk.
 func NewDocumentDecoder(r io.ReadCloser) io.ReadCloser {
 	scanner := bufio.NewScanner(r)
+	// Adding maxCapacity to set the Buffer size to 5M to fix #89016
+	maxCapacity := 5120 * 1024
+	buf := make([]byte, maxCapacity)
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go]: this allocates 5MB of memory on every yaml document read? That seems less than optimal. 

Is there an alternative method that could stream/split without preallocating large chunks like this?

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go]: From the scanner docs:

> Programs that need more control over error handling or large tokens, or must run sequential scans on a reader, should use bufio.Reader instead.

That would be more what I expected

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go]: alternatively, it looks like you can pass a smaller buffer and a larger max size, and the scanner will grow the buffer up to the max size if required. That would be an equally minimal fix and more acceptable. 

This also needs an accompanying test demonstrating the issue and fix.

Comment (tanjunchen) [reply staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go]:
```
scanner := bufio.NewScanner(r)
	buf := make([]byte, 64*1024)
	scanner.Buffer(buf, 5*1024*1024)
	scanner.Split(splitYAMLDocument)
	return &YAMLDecoder{
		r:       r,
		scanner: scanner,
	}
```
how about it?How big is the buffer set? What about the maximum?

Comment (0xMH) [reply staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go]: > this allocates 5MB of memory on every yaml document read? 

No I don't think this is what happens. So from the [Buffer SourceCode](https://golang.org/src/bufio/scan.go?s=9061:9106#L252)  it assigns the initial s.buf to cap(buf)
```
s.buf = buf[0:cap(buf)]
s.maxTokenSize = max
```
and then on Scan() It starts from s.buf and builds a buffer size up to maxTokenSize [SourceCode](https://github.com/golang/go/blob/master/src/bufio/scan.go#L188)
```
newSize := len(s.buf) * 2
if newSize == 0 {
    newSize = startBufSize
}
if newSize > s.maxTokenSize {
    newSize = s.maxTokenSize
}
```

So from my understanding, it's not assigning 5MB of memory to every yaml. it assigns it Gradually for the large documents that need it. Please correct me if I misunderstood something.

Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go]:
```
buf := make([]byte, maxCapacity)
``` 
This line allocates 5MB of memory, regardless of what scanner does

Comment (0xMH) [reply staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go]: Oh I missed that. Changed it now. like @tanjunchen suggested

File: staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go [hunk 2]
```diff
@@ -92,6 +92,8 @@ type YAMLDecoder struct {
 // the caller in framing the chunk.
 func NewDocumentDecoder(r io.ReadCloser) io.ReadCloser {
 	scanner := bufio.NewScanner(r)
+	buf := make([]byte, 64*1024)
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go]: suggest 4k to match existing default buffer size

File: staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go [hunk 3]
```diff
@@ -92,6 +92,8 @@ type YAMLDecoder struct {
...
+	scanner.Buffer(buf, 5*1024*1024)
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go]: how was 5MB selected as a target?

**Comment by liggitt** [comment]
> This also needs an accompanying test demonstrating the issue and fix.

this is still outstanding as well

File: staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go (hunk 3 above)
Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go]: https://github.com/kubernetes/kubernetes/issues/89016#issuecomment-597340082

Random thought from me. My thinking was this is generally client-side, and done ~once per client utility run.

But I agree with the suggestion of a smallish buffer paired with a large max size, that's better.

**Comment by lavalamp** [comment]
+1 for a test.

File: staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go (hunk 2 above)
Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/util/yaml/decoder.go]: +1

**Comment by tanjunchen** [comment]
> > This also needs an accompanying test demonstrating the issue and fix.
> 
> this is still outstanding as well

How to test this change? Any good suggestions? I can learn. I have a problem with this  https://github.com/kubernetes/kubernetes/compare/master...tanjunchen:Buffer-show?expand=1#diff-3b3ea800b8ebe625d0defd76c2e6f827R75. Any suggestions? Thanks /cc @liggitt

**Comment by lavalamp** [comment]
> How to test this change? Any good suggestions?

Generate a really big YAML file and see if this function will handle it. Maybe use `strings.Repeat()` to make sufficiently big items. (look in `*_test.go` files for examples)

**Comment by tanjunchen** [comment]
> > How to test this change? Any good suggestions?
> 
> Generate a really big YAML file and see if this function will handle it. Maybe use `strings.Repeat()` to make sufficiently big items. (look in `*_test.go` files for examples)

```
func TestLongYAML(t *testing.T) {
	d := `
stuff: 1
`
	MaxLen := 5 * 1024 * 1024
	bufferLen := 4 * 1024
	//  MaxLen 5242880  len(dd) 5242880
	dd := strings.Repeat(d, 512*1024)
	r := NewDocumentDecoder(ioutil.NopCloser(bytes.NewReader([]byte(dd[:MaxLen-1]))))
	b := make([]byte, bufferLen)
	n, err := r.Read(b)
	if err != io.ErrShortBuffer{
		t.Fatalf("expected ErrShortBuffer: %d / %v", n, err)
	}
	b = make([]byte, MaxLen)
	n, err = r.Read(b)
	if err != nil {
		t.Fatalf("expected nil: %d / %v", n, err)
	}
	r = NewDocumentDecoder(ioutil.NopCloser(bytes.NewReader([]byte(dd))))
	b = make([]byte, MaxLen)
	n, err = r.Read(b)
	if err != bufio.ErrTooLong {
		t.Fatalf("bufio.Scanner: token too long: %d / %v", n, err)
	}
}
```
/cc  @lavalamp

**Comment by lavalamp** [comment]
Right, that looks generally like I expect :)

**Comment by 0xMH** [comment]
@lavalamp So Should I add @tanjunchen 's test case?

**Comment by lavalamp** [comment]
Well, this needs some sort of test...

**Comment by 0xMH** [comment]
> Well, this needs some sort of test...

Can anyone provide some specific info about the needed tests and changes so I can learn and implement it?.

**Comment by lavalamp** [comment]
The purpose of the test is to:
a) prove that the PR does what we think it does
b) ensure that the property we want (of this working on large documents) continues to exist even after many other people change this code.

Also, tests c) should be repeatable (low--ideally, zero--false positive / false negative rate).

Writing a test is usually more than 50% of the work of making a change, but especially in big projects, it's mandatory or things quickly regress.

Does that help?

**Comment by tanjunchen** [comment]
@0xMH 
any progress?

**Comment by 0xMH** [comment]
Sorry I don't have much time on my hand these days. I guess someone else should take it.
